lib/toaster/bldcontrol/localhostbecontroller.py

```diff
@@ -0,0 +1,516 @@
+#
+# BitBake Toaster Implementation
+#
+# Copyright (C) 2014        Intel Corporation
+#
+# SPDX-License-Identifier: GPL-2.0-only
+#
+
+import os
+import re
+import shutil
+import time
+from bldcontrol.models import BuildEnvironment, BuildRequest, Build
+from orm.models import CustomImageRecipe, Layer, Layer_Version, Project, ToasterSetting
+from orm.models import signal_runbuilds
+import subprocess
+
+from toastermain import settings
+
+from bldcontrol.bbcontroller import BuildEnvironmentController, ShellCmdException, BuildSetupException
+
+import logging
+logger = logging.getLogger("toaster")
+
+install_dir = os.environ.get('TOASTER_DIR')
+
+from pprint import pformat
+
+class LocalhostBEController(BuildEnvironmentController):
+    """ Implementation of the BuildEnvironmentController for the localhost;
+        this controller manages the default build directory,
+        the server setup and system start and stop for the localhost-type build environment
+
+    """
+
+    def __init__(self, be):
+        super(LocalhostBEController, self).__init__(be)
+        self.pokydirname = None
+        self.islayerset = False
+
+    def _shellcmd(self, command, cwd=None, nowait=False,env=None):
+        if cwd is None:
+            cwd = self.be.sourcedir
+        if env is None:
+            env=os.environ.copy()
+
+        logger.debug("lbc_shellcmd: (%s) %s" % (cwd, command))
+        p = subprocess.Popen(command, cwd = cwd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=env)
+        if nowait:
+            return
+        (out,err) = p.communicate()
+        p.wait()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Invoking `Popen.wait()` with `stdout=PIPE` or `stderr=PIPE` might lead to a deadlock if the child process generates enough output to block the OS pipe buffer and waits for it to accept more data. We recommend replacing this usage with `Popen.communicate()` API. In case of asynchronous wait, consider using `asyncio` module APIs.

[module-asyncio](https://docs.python.org/3/library/asyncio.html#module-asyncio)
[Learn more](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait)